python/mxnet/optimizer.py

```diff
@@ -650,6 +651,57 @@ def update(self, index, weight, grad, state):
         weight[:] -= current_delta + wd * weight
 
 @register
+class Ftrl(Optimizer):
+    """
+    Reference:Ad Click Prediction: a View from the Trenches
+
+    Parameters
+    ----------
+    lamda1 : float, optional
+        L1 regularization coefficient.
+
+    learning_rate : float, optional
+        The initial learning rate.
+
+    beta : float, optional
+        Per-coordinate learning rate correlation parameter.
+    eta_{t,i}=frac{learning_rate}{beta+sqrt{sum_{s=1^}tg_{s,i}^t}
+
+    """
+
+    def __init__(self, lamda1=0.01, learning_rate=0.1, beta=1, **kwargs):
+        super(Ftrl, self).__init__(**kwargs)
+        self.lamda1 = lamda1
+        self.beta = beta
+        self.lr = learning_rate
+
+    def create_state(self, index, weight):
+        return (zeros(weight.shape, weight.context),  # dn
+                zeros(weight.shape, weight.context))  # n
+
+    def update(self, index, weight, grad, state):
+        assert(isinstance(weight, NDArray))
+        assert(isinstance(grad, NDArray))
+        self._update_count(index)
+        wd = self._get_wd(index)
+
+        # preprocess grad
+        grad *= self.rescale_grad
+        if self.clip_gradient is not None:
+            grad = clip(grad, -self.clip_gradient, self.clip_gradient)
+
+        # accumulated g and delta initlization
+        dn, n = state
+
+        #update dn, n
+        dn[:] += grad - (sqrt(n + grad * grad) - sqrt(n)) * weight / self.lr
+        n[:] += grad * grad
```

Comment:
We can use `dn += grad - (sqrt(n + grad * grad) - sqrt(n)) * weight / self.lr` and `n += grad * grad` directly. `+=` is inplace.